cranelift-wasm/src/code_translator.rs

```diff
@@ -987,18 +987,86 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
         | Operator::Fence { .. } => {
             return Err(wasm_unsupported!("proposed thread operator {:?}", op));
         }
-        Operator::MemoryInit { .. }
-        | Operator::DataDrop { .. }
-        | Operator::MemoryCopy
-        | Operator::MemoryFill
-        | Operator::TableInit { .. }
-        | Operator::ElemDrop { .. }
-        | Operator::TableCopy
-        | Operator::TableGet { .. }
-        | Operator::TableSet { .. }
-        | Operator::TableGrow { .. }
-        | Operator::TableSize { .. } => {
-            return Err(wasm_unsupported!("proposed bulk memory operator {:?}", op));
+        Operator::MemoryCopy => {
+            // The WebAssembly MVP only supports one linear memory and we assume it here.
+            let heap_index = MemoryIndex::from_u32(0);
+            let heap = state.get_heap(builder.func, 0, environ)?;
+            let len = state.pop1();
+            let src = state.pop1();
+            let dest = state.pop1();
+            environ.translate_memory_copy(builder.cursor(), heap_index, heap, dest, src, len)?;
+        }
+        Operator::MemoryFill => {
+            // The WebAssembly MVP only supports one linear memory and we assume it here.
+            let heap_index = MemoryIndex::from_u32(0);
+            let heap = state.get_heap(builder.func, 0, environ)?;
+            let len = state.pop1();
+            let val = state.pop1();
+            let dest = state.pop1();
+            environ.translate_memory_fill(builder.cursor(), heap_index, heap, dest, val, len)?;
+        }
+        Operator::MemoryInit { segment } => {
+            // The WebAssembly MVP only supports one linear memory and we assume it here.
+            let heap_index = MemoryIndex::from_u32(0);
+            let heap = state.get_heap(builder.func, 0, environ)?;
+            let len = state.pop1();
+            let src = state.pop1();
+            let dest = state.pop1();
+            environ.translate_memory_init(
+                builder.cursor(),
+                heap_index,
+                heap,
+                *segment,
+                dest,
+                src,
+                len,
+            )?;
+        }
+        Operator::DataDrop { segment } => {
+            environ.translate_data_drop(builder.cursor(), *segment)?;
+        }
+        Operator::TableSize { table } => {
+            state.push1(environ.translate_table_size(builder.cursor(), *table)?);
```

Comment:
You could use the following instead:

```suggestion
            let table_data = &builder.cursor().func.tables[*table];
            let base = builder.cursor().ins().global_value(table_data.index_type, table_data.base_gv);
            let bound = builder.cursor().ins().global_value(table_data.index_type, table_data.bound_gv);
            state.push1(builder.cursor().ins().isub(bound, base));
```

Maybe even add an instruction for it, just like `table_addr`? (legalized at https://github.com/bytecodealliance/cranelift/blob/3317fc623c315073710fc1ed67c2d17d2466ab5d/cranelift-codegen/src/legalizer/table.rs#L14)

Reply:
I'll take a look at doing this.

Reply:
It seems like this would work, however on further thought, I think that we should leave this to the environ as well for consistency. If we could implement most of the bulk-memory operators in an embedding neutral way then I wouldn't mind. But if this is the only one, it seems like there could be some confusion.